kong/db/declarative/init.lua

```diff
@@ -798,174 +796,91 @@ function declarative.load_into_cache(entities, meta, hash, shadow)
     -- tags:admin|@list -> all tags tagged "admin", regardless of the entity type
     -- each tag is encoded as a string with the format "admin|services|uuid", where uuid is the service uuid
     local key = "tags:" .. tag_name .. "|@list"
-    local ok, err = kong.core_cache:safe_set(key, tags, shadow)
+    local tags, err = marshall(tags)
+    if not tags then
+      return nil, err
+    end
+
+    local ok, err = t:set(key, tags)
     if not ok then
       return nil, err
     end
   end
 
   -- tags||@list -> all tags, with no distinction of tag name or entity type.
   -- each tag is encoded as a string with the format "admin|services|uuid", where uuid is the service uuid
-  local ok, err = kong.core_cache:safe_set("tags||@list", tags, shadow)
-  if not ok then
+  local tags, err = marshall(tags)
+  if not tags then
     return nil, err
   end
 
-  local ok, err = ngx.shared.kong:safe_set(DECLARATIVE_HASH_KEY, hash or true)
+  t:set("tags||@list", tags)
+
+  hash = hash or "true"
+  t:set(DECLARATIVE_HASH_KEY, hash)
+
+  kong.default_workspace = default_workspace
+
+  local ok, err = t:commit()
   if not ok then
-    return nil, "failed to set " .. DECLARATIVE_HASH_KEY .. " in shm: " .. err
+    return nil, err
   end
 
+  kong.core_cache:purge()
+  kong.cache:purge()
```

Comment:
Maybe purge after the router rebuild.

Reply:
I have reviewed my concerns about updates with Datong.

The TLDR is: we *might* add a block *at the beginning of the procedure* (instead of just blocking the page access, like we did before, we might block the whole thing, including JSON parsing) on the final implementation. LMDB updates are faster than our previous system and the transactions take care of the "multiple page" problem so it might not be necessary, it might require some tests.